packages/block-library/src/tabs/view.js

```diff
@@ -0,0 +1,212 @@
+/**
+ * WordPress dependencies
+ */
+import { store, getContext, getElement } from '@wordpress/interactivity';
+
+/**
+ * Determines whether a value is numeric.
+ *
+ * @param {*} value A vlue to check.
+ * @return {boolean} Whether the value is numeric.
+ */
+function isNumeric( value ) {
+	return ! isNaN( parseFloat( value ) );
+}
+
+/**
+ * Generates a map of tab panel IDs to tab label IDs.
+ *
+ * @param {Element[]} tabLabels List of tab label elements.
+ * @return {Map} A map of tab panel IDs to tab label IDs.
+ */
+function getTabPanelToLabelIdMap( tabLabels ) {
+	if ( ! tabLabels ) {
+		return new Map();
+	}
+
+	return new Map(
+		tabLabels.map( ( tabLabel ) => {
+			const tabPanelId = tabLabel.getAttribute( 'href' )?.substring( 1 );
+			const tabLabelId = tabLabel.getAttribute( 'id' );
+			return [ tabPanelId, tabLabelId ];
+		} )
+	);
+}
+
+/**
+ * Initializes the tabs block with the necessary ARIA attributes.
+ *
+ * @param {HTMLElement} ref The block wrapper element.
+ */
+function initTabs( ref ) {
```

Comment:
What's the reason for doing all these DOM manipulations here manually instead of using directives for it? 

This way they won't get server rendered and there will be a flash of unsettled content as the page loads 🤔

Reply:
@fabiankaegy and @luisherranz Thank you for the feedback!

Right now, by default, the tabs block is a list of on page links, and then progressively enhanced into tabs when JS is loaded.

How I can add the attributes on the server and still progressively enhance the block depending on if JS is loaded or not? Is there a way to use the interactivity API to add the role and aria attributes conditionally?

Reply:
> Is there a way to use the interactivity API to add the role and aria attributes conditionally?

Perhaps the guides that will be published here can help you understand how the Interactivity API works? 

> This is not published yet, but maybe [this guide](https://github.com/WordPress/gutenberg/pull/63759), which is intended to explain this declarative approach, can help (I'd also appreciate any feedback regarding the guide itself 🙂).

I know I'm repeating myself, but I think the accordion block example is really helpful for understanding how to program something like this, since implementing an accordion is quite similar to creating tabs.

> Here's a [good example of a block](https://github.com/WordPress/gutenberg/pull/64119) whose functionality is similar to this one. Take a special look at [the frontend file](https://github.com/WordPress/gutenberg/pull/64119/files#diff-3d35108b849bece54985a937368eaa12364bb6f97c17d08683541b6eeb38ce36): it only modifies the local context. All the DOM manipulations are done through directives that are linked to that local context. Also, notice how the local context is initialized on the server to ensure everything is ready before the JS downloads and executes.

Reply:
> This is not published yet, but maybe this guide, which is intended to explain this declarative approach, can help (I'd also appreciate any feedback regarding the guide itself 🙂).

The guide was published [here](https://developer.wordpress.org/block-editor/reference-guides/interactivity-api/core-concepts/the-reactive-and-declarative-mindset/).

Reply:
@fabiankaegy, @luisherranz, @DAreRodz In https://github.com/WordPress/gutenberg/pull/63689/commits/cc304b55ad47d7b63985e39d965f420023c2f44d, I've removed the `initTabs` callback in favor of server-side directives and state getters. This is less code overall, so thank you for the guidance.

One thing clarify is that the tabs front-end here has been built as progressively enhanced markup, as suggested by the likes of https://design-system.service.gov.uk/components/tabs/ and https://inclusive-components.design/tabbed-interfaces/. Without JS, it's simply a list of on page links:

![Screen Recording 2024-08-29 at 11 21 01](https://github.com/user-attachments/assets/223e3f59-6487-4cf4-8f52-74e78de8ca29)

This means that I'm not adding attributes directly with the interactivity API (`$p->set_attribute( 'role', 'tab' );`), but instead using the `wp-bind` directive (`$p->set_attribute( 'data-wp-bind--role', 'state.roleAttribute' );`) with a getter in the `view.js` state that returns the correct role based on the element's class name. The `role`, `tabindex`, and `aria-` attributes are semantically incorrect if JS is not loaded on the page, as there is no tabbed interface, so I don't think they should be present directly in the server-side markup.

For attributes with dynamic values this works well. It would also be nice to have a way to do this more simply with static values. For example, `$p->set_attribute( 'data-wp-bind--role', 'tablist` )` doesn't work right now, but this would simplify the code even more, as the role property never changes, it just needs to be added when JS is initialized.